scarlet/wavelet.py

```diff
@@ -0,0 +1,300 @@
+import autograd.numpy as np
+from . import fft
+from . import Cache
+from scipy.stats import median_absolute_deviation as mad
+
+# Filter for the scarlet transform. Here bspline
+h = np.array([1. / 16, 1. / 4, 3. / 8, 1. / 4, 1. / 16])
+
+class Starlet(object):
+    """ A class used to create the Wavelet transform of a cube of images from the 'a trou' algorithm.
+
+        The transform is performed by convolving the image by a seed starlet: the transform of an all-zero
+        image with its central pixel set to one. This requires 2-fold padding of the image and an odd pad
+        shape. The fft of the seed starlet is cached so that it can be reused in the transform of other
+        images that have the same shape.
+    """
+    def __init__(self, image, lvl = None, starlet = None, direct = False):
+        """ Initialise the Starlet object
+
+        Paramters
+        image: Fourier object
+            image to transform
+        lvl: int
+            number of starlet levels to use in the decomposition
+        starlet: array
+            Starlet transform of an array
```

Comment:
the transform of *which* array? I'm guessing of `image`...

Reply:
No, because Starlet objects can be instanciated with their transform, so as an argument, it can never be the transform of `image`.

Reply:
I'm not sure I understand your logic. It looks like if `starlet` is not set, the init will compute it from `image`. If it is set, init won't compute it (fine), but then it should still be the transform of `image`.

Reply:
I see the trouble. A Starlet object is a container for an image and its transform. It can either be instantiated from the image or from the starlet coefficients. If 'image` and 'starlet' are both provided, it is up to the user to make sure they know what they are doing. Alternatively, I could make it so that 'image' and 'starlet' (to be replaced by 'coefficients') are both optional parameters and mutually exclusive. I.e. you can declare a Starlet object from both ends, but for consistency, we avoid being able to provide both.

Reply:
see my suggestion on refactoring below

Reply:
I'm closing this as it is delt with below.